lib/Grape.js

```diff
@@ -217,58 +256,145 @@ class Grape extends Events {
+    const done = !cb ? noop : (err, key) => {
+      if (err) {
+        cb(err)
+        return
+      }
+      cb(null, this.str2hex(key))
     }
+
+    const mutable = 'k' in opts
+    if (mutable) this.mutablePut(opts, done)
+    else this.immutablePut(opts.v, done)
   }
 
   get (opts, cb) {
-    if (typeof opts === 'string') opts = this.str2buf(opts, 'hex')
-    if (Buffer.isBuffer(opts)) opts = { hash: opts }
+    if (typeof opts === 'string' || Buffer.isBuffer(opts)) {
+      this.immutableGet(opts, cb)
+      return
+    }
 
-    if (opts.salt && !Buffer.isBuffer(opts.salt)) {
-      opts.salt = this.str2buf(opts.salt)
+    // any API's using opts.hash for mutable data should
+    // switch to opts.key
+    const mutable = opts.key || opts.m
+
+    // if salt/seq is used with opts.hash we can infer mutable
+    const inferredMutable = !mutable && opts.hash && (opts.salt || 'seq' in opts)
+
+    if (mutable || inferredMutable) {
+      this.mutableGet(opts, cb)
+      return
     }
 
-    try {
-      this.node.get(opts.hash, opts, (err, res) => {
-        if (!res) {
-          return cb(err)
+    // explicitly immutable
+    if (opts.hash && opts.m === false) {
+      this.immutableGet(opts.hash, cb)
+      return
+    }
+    // in this case we can't tell whether the get is immutable
+    // or mutable
+    if (opts.hash) {
+      this.mutableGet(opts, (err, res) => {
+        if (err) {
+          cb(err)
+          return
         }
+        // mutable store didn't have any values, try mutable
+        if (res.v === null) {
+          this.immutableGet(opts.hash, cb)
+        } else {
+          cb(null, res)
+        }
+      })
+      return
+    }
 
-        res.id = this.str2hex(res.id)
-        res.v = res.v.toString()
+    cb(Error('ERR_GRAPE_INVALID_GET_OPTIONS'))
+  }
 
-        if (res.k) {
-          res.k = res.k.toString('hex')
+  immutablePut (value, cb) {
+    try {
+      value = typeof value === 'string' ? this.str2buf(value) : value
+      const qs = this.node.immutable.put(value, cb)
+      qs.on('warning', this[kOnWarning])
+    } catch (e) {
+      const msg = `ERR_GRAPE_GENERIC: ${e.message}\n${e.stack}`
+      cb(new Error(msg))
+    }
+  }
+
+  mutablePut (opts, cb) {
+    try {
+      const { k: publicKey, v, sig, ...options } = opts
+      options.signature = sig
+      options.keypair = { publicKey }
+      if (options.sign) {
+        throw Error('ERR_GRAPE_SIGN_NOT_SUPPORTED')
+      }
+      const value = typeof v === 'string' ? this.str2buf(v) : v
+      const qs = this.node.mutable.put(value, options, (err, { key }) => {
+        cb(err, key)
+      })
+      qs.on('warning', this[kOnWarning])
+    } catch (e) {
+      const msg = `ERR_GRAPE_GENERIC: ${e.message}\n${e.stack}`
+      cb(new Error(msg))
+    }
+  }
+
+  immutableGet (hash, cb) {
+    if (typeof hash === 'string') hash = this.str2buf(hash, 'hex')
+    try {
+      const qs = this.node.immutable.get(hash, (err, value, { id }) => {
```

Comment:
```
(18:26:51) [robert@tequilax1] ~/bitfinex/grenache-nodejs-link/examples (master) $ node put_get_mutable
putMutable: null 0361eef098a0c703a4f4ec8fe1034e8faafe70f6e40f7b4cbb3a8853a36c59ba
data requested from the DHT { Error: socket hang up
    at createHangUpError (_http_client.js:323:15)
    at Socket.socketOnEnd (_http_client.js:426:23)
    at Socket.emit (events.js:203:15)
    at endReadableNT (_stream_readable.js:1145:12)
    at process._tickCallback (internal/process/next_tick.js:63:19) code: 'ECONNRESET' } undefined
^C

```

Reply:
```
(18:26:57) [robert@tequilax1] ~/bitfinex/bfx-ext-feed-js (pr/49 *) $ DEBUG=* grape --dp 20001 --aph 30001 --bn '127.0.0.1:20002'
  grenache:grape starting +0ms
  grenache:grape 20001 listening +6ms
  grenache:grape 20001 ready +3s
  grenache:grape 20001 node +613ms
  grenache:grape 20001 warning { Error: Request timed out
    at Object.<anonymous> (/home/robert/bitfinex/grenache-grape/node_modules/dht-rpc/lib/io.js:10:19)
    at Module._compile (internal/modules/cjs/loader.js:778:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at require (internal/modules/cjs/helpers.js:25:18)
    at Object.<anonymous> (/home/robert/bitfinex/grenache-grape/node_modules/dht-rpc/index.js:10:12)
    at Module._compile (internal/modules/cjs/loader.js:778:30) code: 'ETIMEDOUT' } +9s
/home/robert/bitfinex/grenache-grape/lib/Grape.js:349
      const qs = this.node.immutable.get(hash, (err, value, { id }) => {
                                                              ^

TypeError: Cannot destructure property `id` of 'undefined' or 'null'.
    at qs.node.immutable.get (/home/robert/bitfinex/grenache-grape/lib/Grape.js:349:48)
    at QueryStream.finished (/home/robert/bitfinex/grenache-grape/node_modules/@hyperswarm/dht/stores.js:73:11)
    at QueryStream.f (/home/robert/bitfinex/grenache-grape/node_modules/once/once.js:25:25)
    at QueryStream.onerror (/home/robert/bitfinex/grenache-grape/node_modules/end-of-stream/index.js:44:12)
    at QueryStream.emit (events.js:198:13)
    at QueryStream.destroy (/home/robert/bitfinex/grenache-grape/node_modules/dht-rpc/lib/query-stream.js:247:19)
    at QueryStream._finalize (/home/robert/bitfinex/grenache-grape/node_modules/dht-rpc/lib/query-stream.js:217:12)
    at QueryStream._moveCloser (/home/robert/bitfinex/grenache-grape/node_modules/dht-rpc/lib/query-stream.js:205:12)
    at QueryStream._read (/home/robert/bitfinex/grenache-grape/node_modules/dht-rpc/lib/query-stream.js:235:39)
    at QueryStream._readMaybe (/home/robert/bitfinex/grenache-grape/node_modules/dht-rpc/lib/query-stream.js:227:70)
(18:28:46) [robert@tequilax1] ~/bitfinex/bfx-ext-feed-js (pr/49 *) $ 
```

Reply:
destructuring in function args is always a bit dangerous.

woudl recommend to switch to this pattern in general (even though the bug originates from somethere else):

```js
this.node.immutable.get(hash, (err, value, opts) => {
  const { id } = opts
})
```

Reply:
code used: https://github.com/bitfinexcom/grenache-nodejs-link/blob/master/examples/put_get_mutable.js

Reply:
other grape node says:

```
buffer.js:515
    throw new ERR_INVALID_ARG_TYPE(
    ^

TypeError [ERR_INVALID_ARG_TYPE]: The "string" argument must be one of type string, Buffer, or ArrayBuffer. Received type object
    at Function.byteLength (buffer.js:515:11)
    at bufferLength (/home/robert/bitfinex/grenache-grape/node_modules/protocol-buffers-encodings/index.js:285:53)
    at Object.encodingLength (/home/robert/bitfinex/grenache-grape/node_modules/protocol-buffers-encodings/index.js:66:15)
    at encodingLength (/home/robert/bitfinex/grenache-grape/node_modules/dht-rpc/lib/messages.js:154:24)
    at Object.encode (/home/robert/bitfinex/grenache-grape/node_modules/dht-rpc/lib/messages.js:162:40)
    at IO.response (/home/robert/bitfinex/grenache-grape/node_modules/dht-rpc/lib/io.js:231:23)
    at reply (/home/robert/bitfinex/grenache-grape/node_modules/dht-rpc/index.js:158:16)
    at Object.query (/home/robert/bitfinex/grenache-grape/node_modules/@hyperswarm/dht/stores.js:123:9)
    at HyperDHT._oncommand (/home/robert/bitfinex/grenache-grape/node_modules/dht-rpc/index.js:153:14)
    at HyperDHT.onrequest (/home/robert/bitfinex/grenache-grape/node_modules/dht-rpc/index.js:98:21)

```

which then leads to the request timing out ...

Reply:
when the request times out, i guess nodes should not crash....

Reply:
I've covered the error handling in get, and I've made the error output more useful
we should get a convo about error handling in `Link` at some point, it seems like there's cases where errors are sent as responses instead of propagated as errors

with regard to the failure, I can reproduce this if I provide a bootstrap node that isn't running, but if I provide one that is then it does work

the node will respond (without crashing now) with the error - "ERR_GRAPE_GENERIC: No nodes responded"